Fixing InboxScenari for PKCE with CAS case

Arsnael opened the pull request:
Well I kinda manage to login / logout with keycloak + cas, and the rest seems ok.

Things left:
* load assets: most of them are commented now because I decided to fix those at the end. While I understand they are part of the simulation  to get something closer to reality, it makes the simulation completely hellish to maintain... Every time you change assets in openpaas or the cas, you have to update the simulation...
* login: there is a slight flow difference with keycloak if it is the first connection with the user or not with keycloak. And not all users have logged in yet. I will try to find a solution to differentiate those cases... Or I could just run the first part of the simulation (login/logout) on the 20000 tests users so we don't care... Will see
* when simulating a user searching for other users and checking their avatar, some users seem to have an avatar with a not found image: https://openpaas.upn.integration-open-paas.org/images/not_a_user.png. Didn't take the time to dig on that yet 

Of course I will need to do a bit of cleanup obviously when I finish to fix everything...

Arsnael commented:
BTW forgot to post yesterday but that was the gatling simulation at the end of the day, for 10 users (1 minute injection + 3 minutes scenario duration):

![screenshot--2021 07 02-14_33_32](https://user-images.githubusercontent.com/9005025/124238339-a9eae480-db42-11eb-875e-029b0df1c4df.png)

chibenwa commented:
We still have some asset calls to fix, mostly, right?

(I can see 48 `404`...)

chibenwa commented:
@Arsnael  Can you post the ZIP of the simulation so that frontend devs can have a look at it?

Arsnael commented:
Something I find really odd in that simulation (gatling inbox): we are in a loop where the user always logs in, then do a bunch of stuff (20 times), then logout... pause between 7 and 15s and repeat the loop... hmmm honestly is that really the reality? I don't think users will keep login/logout like crazy
I'm not surprised that lemonldap had a hard time to keep up before with such a simulation honestly...

should we:
* keep the user login in and out on regular basis but maybe more longer period between?
* just login the user before the loop, logout after it?

Arsnael commented:
> (I can see 48 404...)
@chibenwa those 404 errors are related to my third bullet point: 
```
when simulating a user searching for other users and checking their avatar, some users seem to have an avatar with a not found image: https://openpaas.upn.integration-open-paas.org/images/not_a_user.png. Didn't take the time to dig on that yet
```

Arsnael commented:
The one from today after I:
* fixed the first point with the first login flow case in keycloak
* added back some assets (not everything yet)

[inboxwithrefreshtokensimulation.zip](https://github.com/linagora/openpaas-gatling/files/6754055/inboxwithrefreshtokensimulation.zip)

tanandy commented:
> Something I find really odd in that simulation (gatling inbox): we are in a loop where the user always logs in, then do a bunch of stuff (20 times), then logout... pause between 7 and 15s and repeat the loop... hmmm honestly is that really the reality? I don't think users will keep login/logout like crazy
> I'm not surprised that lemonldap had a hard time to keep up before with such a simulation honestly...
> 
> should we:
> 
> * keep the user login in and out on regular basis but maybe more longer period between?
> * just login the user before the loop, logout after it?

interesting comment, i would like to hear more from @remk and @rouazana

Arsnael commented:
Also other remark, some assets are usually cached in the browser... but we still load them endlessly in those simulations?

Arsnael commented:
kinda fixed the simulation. However, as I commented through the code, some stuff are wrong/missing, and I will try to change that:
* I don't think login/logout should be in a loop here. Ok in fact it seems the user spends maybe at least around 20 minutes before logout/login, so no big deal(ish) but still... it makes the simulation very noisy. If we want to test the load capacity on keycloak / CAS (login/logout), we could just easily make a simulation that does just login/logout in a loop, and see where is the limit.
* I was surprised to notice there is no `getMessages` jmap call after `getMessagesList`. That is something used all the time, need to add this call
* probably some flows  have been optimized since this simulation was written, there is not that many duplicated calls around now
* need to think more of the possible actions of an user in inbox. The idle here for example is not really idle.

Arsnael commented:
Extra question though: not sure to completely get why we don't load static assets from nginx in the simulation? it's not openpaas job anymore ok but it's still being loaded by the user though?

Arsnael commented:
10 users on my local, 1 min injection, 3 minutes scenario run:

![screenshot--2021 07 07-17_32_00](https://user-images.githubusercontent.com/9005025/124745242-ce333080-df49-11eb-99df-8f80c6b0759d.png)

zip: [inboxPlatformValidationLocal.zip](https://github.com/linagora/openpaas-gatling/files/6776430/inboxPlatformValidationLocal.zip)

Arsnael commented:
I think I'm ok with the status now. Notice that:

* regarding static assets, maybe sometimes we want to run that IDK, so I made a conf for it. Of course it's disabled by default (so no load static assets)
* I included getMailboxes in idle, even if... well it seems separated from the 1min sync of getMessagesList + getMessages in the current mailbox. But I'm not sure how to parallelize different chain of actions in gatling (or if it's possible in the first place)
* there is a pause of 60 seconds between each action, as it is the time range sync for idle. But not sure it's really correct...
* I put in the randomSwitch those values:
  * 70% idle
  * 15% read email
  * 10% open a mailbox
  * 5% write an email

Any feedback and discussion is welcomed regarding those points (or anything else), as I'm not 100% sure of the normal usage of inbox by users

chibenwa commented on src/main/scala/com/linagora/openpaas/gatling/unifiedinbox/scenari/InboxScenari.scala:
```diff
@@ -93,9 +93,10 @@ object InboxScenari {
-      .exec(getIdleMessageList)
-      .exec(getMessages))
+    randomSwitch(
+      10.0 -> exec(getMailboxes),
```
```suggestion
      10.0 -> exec(getMailboxes)
        .exec(getIdleMessageList)
        .exec(getMessages))),
```

chibenwa commented on src/test/resources/logback-test.xml.debug:
```diff
@@ -0,0 +1,29 @@
+<configuration>
```
> logback-test.debug.xml

so that we have syntax highlighting

chibenwa commented on src/main/scala/com/linagora/openpaas/gatling/unifiedinbox/scenari/InboxScenari.scala:
```diff
@@ -21,7 +21,6 @@ object InboxScenari {
   def platform(feederBuilder: SourceFeederBuilder[String]) = scenario("Inbox platform scenario")
     .feed(feederBuilder.circular)
       .exec(InboxScenari.userLogin())
-      .exec(getProfile())
```
Why?

Arsnael commented:
4000 users inbox after scaling up a bit and adding cpu to backend-esn and sabre:

![screenshot--2021 07 13-14_24_24](https://user-images.githubusercontent.com/9005025/125409907-9105ed00-e3e6-11eb-8191-92a48277040c.png)

zip: [4000usersInbox.zip](https://github.com/linagora/openpaas-gatling/files/6806768/4000usersInbox.zip)

I think we can merge this. Can somebody help me? I don't have the rights to do it myself

rouazana commented:
> I don't have the rights to do it myself

You should have them now :)

Arsnael commented:
> You should have them now :)

Thanks ;)
